units/src/block.rs

```diff
@@ -0,0 +1,266 @@
+        absolute::Height::from_consensus(h.to_u32())
+    }
+}
+
+/// The block interval.
+///
+/// Block interval is an integer type denoting the number of blocks that has passed since some point
+/// i.e., this type is meant for usage as a relative block measure.
+///
+/// This type is not meant for constructing relative height based timelocks, this is a general
+/// purpose block interval abstraction. For locktimes please see [`locktime::relative::Height`].
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+// Public to try and make it really clear that there are no invariants.
+pub struct BlockInterval(pub u32);
+
+impl BlockInterval {
+    /// Block interval 0 i.e., the current block.
+    pub const ZERO: Self = BlockInterval(0);
+
+    /// The minimum block interval (0).
+    pub const MIN: Self = Self::ZERO;
+
+    /// The maximum block interval.
+    pub const MAX: Self = BlockInterval(u32::MAX);
+
+    /// Creates a block interval from a `u32`.
+    // Because From<u32> is not const.
+    pub const fn from_u32(inner: u32) -> Self { Self(inner) }
+
+    /// Returns block interval as a `u32`.
+    // Because type inference doesn't always work using `Into`.
+    pub const fn to_u32(&self) -> u32 { self.0 }
+}
+
+impl fmt::Display for BlockInterval {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { fmt::Display::fmt(&self.0, f) }
+}
+
+crate::impl_parse_str_from_int_infallible!(BlockInterval, u32, from);
+
+impl From<u32> for BlockInterval {
+    fn from(inner: u32) -> Self { Self::from_u32(inner) }
+}
+
+impl From<BlockInterval> for u32 {
+    fn from(height: BlockInterval) -> Self { height.to_u32() }
+}
+
+impl From<relative::Height> for BlockInterval {
+    /// Converts [`locktime::relative::Height`] to a [`BlockInterval`].
+    ///
+    /// A relative locktime block height has a maximum value of `u16::MAX` where as a
+    /// `BlockInterval` is a thin wrapper around a `u32`, the two types are not interchangeable.
+    fn from(h: relative::Height) -> Self { Self::from_u32(h.value().into()) }
+}
+
+impl TryFrom<BlockInterval> for relative::Height {
+    type Error = TooBigForRelativeBlockHeightError;
+
+    /// Converts [`BlockInterval`] to a [`locktime::relative::Height`].
+    ///
+    /// A relative locktime block height has a maximum value of `u16::MAX` where as a
+    /// `BlockInterval` is a thin wrapper around a `u32`, the two types are not interchangeable.
+    fn try_from(h: BlockInterval) -> Result<Self, Self::Error> {
+        let h = h.to_u32();
+        if h > u16::MAX as u32 {
+            return Err(TooBigForRelativeBlockHeightError(h));
+        }
+        Ok(relative::Height::from(h as u16)) // Cast ok, value checked above
+    }
+}
+
+/// The block interval is too big to be used as a relative lock time.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct TooBigForRelativeBlockHeightError(u32);
```

Comment:
This looks like one more error type. Why not use the same one that conversion from `u32` uses?

Reply:
I'm pushing for the errors to be:
- brain dead simple
- really specific
- uniform and easy to read
- only returned from places that fit _exactly_ to reduce maintenance burden

Using an `absolute::ConversionErrror` violates these principles. Adding an extra error is much better IMO.

Reply:
You mean `relative`? Conversion from `u32` and from `BlockInterval` are effectively the same thing. I don't think what you wrote trumps not having more than one type for the same thing.